service/history/historyEngine.go

```diff
@@ -236,27 +234,37 @@ func (e *historyEngineImpl) registerDomainFailoverCallback() {
 	// first set the failover callback
 	e.shard.GetDomainCache().RegisterDomainChangeCallback(
 		e.shard.GetShardID(),
-		e.shard.GetDomainCache().GetDomainNotificationVersion(),
-		// before the domain change, this will be invoked when (most of time) domain cache is locked
-		func(prevDomain *cache.DomainCacheEntry, nextDomain *cache.DomainCacheEntry) {
-			e.logger.Infof("Domain Change Event: Shard: %v, Domain: %v, ID: %v, Failover Notification Version: %v, Active Cluster: %v, Shard Domain Notification Version: %v\n",
-				e.shard.GetShardID(), nextDomain.GetInfo().Name, nextDomain.GetInfo().ID,
-				nextDomain.GetFailoverNotificationVersion(), nextDomain.GetReplicationConfig().ActiveClusterName, e.shard.GetDomainNotificationVersion())
-
-			failoverPredicate(nextDomain, func() {
-				e.logger.Infof("Domain Failover Start: Shard: %v, Domain: %v, ID: %v\n",
-					e.shard.GetShardID(), nextDomain.GetInfo().Name, nextDomain.GetInfo().ID)
-
-				domainID := nextDomain.GetInfo().ID
-				e.txProcessor.FailoverDomain(domainID)
-				e.timerProcessor.FailoverDomain(domainID)
-			})
+		e.shard.GetDomainNotificationVersion(),
+		func() {
+			e.txProcessor.LockTaskPrrocessing()
+			e.timerProcessor.LockTaskPrrocessing()
 		},
-		// after the domain change, this will be invoked when domain cache is NOT locked
-		func(prevDomain *cache.DomainCacheEntry, nextDomain *cache.DomainCacheEntry) {
-			failoverPredicate(nextDomain, func() {
-				e.logger.Infof("Domain Failover Notify Active: Shard: %v, Domain: %v, ID: %v\n",
-					e.shard.GetShardID(), nextDomain.GetInfo().Name, nextDomain.GetInfo().ID)
+		func(prevDomains []*cache.DomainCacheEntry, nextDomains []*cache.DomainCacheEntry) {
+			defer func() {
+				e.txProcessor.UnlockTaskPrrocessing()
```

Comment:
It will be good to reduce the lock on processors, for the duration of taking the snapshot and creation of failover cursors.

Reply:
checked the start of the failover logic, the logic is pretty light, so it should not take much time